3.1 record format migration

MishaDemianenko opened the pull request:
Update high limit format store version to `vE.H.1`.
Update store migrator to perform migrations of store files **only** in case when format capabilities are different.
Allow to load and use 3.0 high limit format (without fixed references part) as independent HighLimitV3_0 record format.

changelog: Update store version, allow db migrator to skip migration of compatible formats during upgrade.

MishaDemianenko committed:
Allow to load and use 3.0 high limit format (without fixed references part) as independent HighLimitV3_0 record format.
Update high limit format store version to `vE.H.1`.

tinwelint commented on enterprise/kernel/src/test/java/org/neo4j/kernel/impl/store/format/highlimit/HighLimitStoreMigrationTest.java:
```diff
@@ -0,0 +1,101 @@
+    @Test
+    public void haveSameFormatCapabilitiesAsHighLimit3_0()
+    {
+        HighLimit.RECORD_FORMATS.hasSameCapabilities( HighLimitV3_0.RECORD_FORMATS, CapabilityType.FORMAT );
```
isn't there an assertion missing here?

tinwelint commented on enterprise/kernel/src/main/java/org/neo4j/kernel/impl/store/format/highlimit/v30/HighLimitV3_0.java:
```diff
@@ -55,7 +55,8 @@
     {
-        super( STORE_VERSION, 7, Capability.DENSE_NODES, Capability.SCHEMA, Capability.LUCENE_5 );
+        super( STORE_VERSION, 7, Capability.DENSE_NODES, Capability.SCHEMA, Capability.LUCENE_5,
+                Capability.FIXED_REFERENCE );
```
Was this added as a hack to not require a migration from vE.H.0?

MishaDemianenko replied:
Correct, if we need to perform store migration only when store capabilities changed - we need new capability for newly introduced fixed references - to avoid migration between high limit 3.0 and high limit 3.1, but still have migration from standard format.

tinwelint replied:
The condition in `StoreMigrator` was changed from `fromFormat.equals( toFormat )` to only comparing format capabilities. I don't think that's what we want because standard and highlimit format has the same _capabilities_, that doesn't mean they look the same on disk. Introducing new format capabilities, like dense nodes f.ex. is to signal that there will need to be restructuring of the records, i.e. a complete migration even if the record bits on disk would be the same. Format capability wasn't intended to be used for cases like this, so what happens if we revert the condition in `StoreMigrator` and just don't introduce the `FIXED_REFERENCE` capability at all?

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/Capability.java:
```diff
@@ -38,6 +38,11 @@
     DENSE_NODES( CapabilityType.FORMAT, CapabilityType.STORE ),
 
     /**
+     * Store has fixed reference encoding support support
```
support support

MishaDemianenko committed:
Update java doc and assertion

MishaDemianenko commented:
@tinwelint updated

MishaDemianenko committed:
Remove introduced Fixed reference capability.
Introduce format family and use it as one of the criteria for format migration.
Based on format family not it's possible to see what kind of format do we see - Standard, High limit or other based on its family that is the same across all versions.

MishaDemianenko commented:
@tinwelint updated

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/FormatFamily.java:
```diff
@@ -0,0 +1,70 @@
+     */
+    public static boolean isUpgradable( FormatFamily oldFamily, FormatFamily newFamily )
+    {
+        return oldFamily.compareTo( newFamily ) < 0;
```
What happens if rank diff is 0? Shouldn't _upgradable_ say _yes_ if the two families are the same? That would also remove the need for the inverted _isNot_ method below.

MishaDemianenko replied:
Not really since format should not be treated as one that need upgrade in case of it just have same family right? only if we migrate some family with younger rank to family that have bigger rank.

tinwelint replied:
Although we can have migration between different versions of format within the same family, take 2.1 -> 2.2 standard format for example... 2.1 is perhaps not supported anymore, but that's beside the point. 2.1 -> 2.2 is of the same format family but have different capabilities and so needs migration.

MishaDemianenko replied:
yes that's absolutely true that's why we have or condition to verify do we need a migration, remember that? lets talk tomorrow should be able to resolve all open questions really fast

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java:
```diff
@@ -170,7 +171,7 @@ public void migrate( File storeDir, File migrationDir, MigrationProgressMonitor.
         RecordFormats oldFormat = selectForVersion( versionToMigrateFrom );
         RecordFormats newFormat = selectForVersion( versionToMigrateTo );
-        if ( !oldFormat.hasSameCapabilities( newFormat, CapabilityType.FORMAT ) )
+        if ( upgradeFormatType( oldFormat, newFormat ) || isDifferentCapabilities( oldFormat, newFormat ) )
```
Is this condition correct? It seems like `isDifferentCapabilities` may override the `false` returned by the `FormatFamily` check.

MishaDemianenko replied:
This condition handle 2 cases - we have different families that we should migrate or we have same families but with different capabilities. Case when we have some not comparable families handled in `UpgradableDatabase`.
Maybe it's more naming thing that brings confusion since those 2 questions are related and highlight confusion around that?

tinwelint commented on integrationtests/logs/http.log:
```diff
@@ -0,0 +1,2 @@
+2016-08-29 17:34:26.028+0000 INFO  [REQUEST] [AsyncLog @ 2016-08-29 17:34:26.027+0000]  null - null [Mon Aug 29 19:34:26 CEST 2016] "/db/data/node?null" 201 -1 "null" "Java/1.8.0_101" 1472492066027
```
Looks like a mistake to include these log changes.

tinwelint commented:
Generally how does the new `FormatFamily#rank()` compare with `RecordFormats#generation()`? Redundant in a way, isn't it?

MishaDemianenko commented:
Family rank allows us to compare different format families - same family preserve it's rank across different versions that allow to compare for example community format and enterprise format in general. So we can determine where this format belongs to - to which group of formats.

MishaDemianenko commented:
In the same time it looks like we using generations to select newest ever format during start up, independently from where it comes from

MishaDemianenko commented:
So to sum up we using generation to compare and see the difference in each particular format implementations and format family rank compares formats families itself: like `community` vs `enterprise`

MishaDemianenko committed:
Remove mistakenly added file.

MishaDemianenko commented:
@tinwelint ping

MishaDemianenko committed:
Make generation format family specific.
Update criteria for format migration, add helper methods in format family class.

MishaDemianenko commented:
@tinwelint updated.
